File: 3414-maximum-score-of-non-overlapping-intervals/3414-maximum-score-of-non-overlapping-intervals.ts
function maximumWeight(intervals: number[][]): number[] {
    const n = intervals.length;
    const K = 4;

    const a: number[][] = intervals.map((x, i) => [x[0], x[1], x[2], i]);

    a.sort((x, y) => x[1] - y[1]);

    const ends: number[] = a.map(x => x[1]);

    type State = {
        score: number;
        ids: number[];
    };

    const dp: Array<Array<State | null>> = Array.from(
        { length: K + 1 },
        () => Array<State | null>(n + 1).fill(null)
    );

    for (let i = 0; i <= n; i++) {
        dp[0][i] = { score: 0, ids: [] };
    }

    const better = (a: State | null, b: State | null): boolean => {
        if (a === null) return false;
        if (b === null) return true;

        if (a.score !== b.score) return a.score > b.score;

        const len = Math.min(a.ids.length, b.ids.length);

        for (let i = 0; i < len; i++) {
            if (a.ids[i] !== b.ids[i]) {
                return a.ids[i] < b.ids[i];
            }
        }

        return a.ids.length < b.ids.length;
    };

    const lowerBound = (length: number, target: number): number => {
        let lo = 0;
        let hi = length;

        while (lo < hi) {
            const mid = lo + Math.floor((hi - lo) / 2);

            if (ends[mid] >= target) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }

        return lo;
    };

    for (let i = 1; i <= n; i++) {
        const left = a[i - 1][0];
        const weight = a[i - 1][2];
        const index = a[i - 1][3];

        const p = lowerBound(i - 1, left);

        for (let k = 1; k <= K; k++) {
            dp[k][i] = dp[k][i - 1];

            const prev = dp[k - 1][p];

            if (prev !== null) {
                const ids = [...prev.ids, index];

                ids.sort((x, y) => x - y);

                const take: State = {
                    score: prev.score + weight,
                    ids
                };

                if (better(take, dp[k][i])) {
                    dp[k][i] = take;
                }
            }
        }
    }

    let answer: State | null = null;

    for (let k = 1; k <= K; k++) {
        if (better(dp[k][n], answer)) {
            answer = dp[k][n];
        }
    }

    return answer!.ids;
}

File: 3414-maximum-score-of-non-overlapping-intervals/3414-maximum-score-of-non-overlapping-intervals.go
func maximumWeight(intervals [][]int) []int {
	type State struct {
		score  int64
		ids    []int
		valid  bool
	}

	n := len(intervals)
	const K = 4

	a := make([][4]int64, n)

	for i := 0; i < n; i++ {
		a[i] = [4]int64{
			int64(intervals[i][0]),
			int64(intervals[i][1]),
			int64(intervals[i][2]),
			int64(i),
		}
	}

	sort.Slice(a, func(i, j int) bool {
		return a[i][1] < a[j][1]
	})

	ends := make([]int64, n)
	for i := 0; i < n; i++ {
		ends[i] = a[i][1]
	}

	dp := make([][]State, K+1)
	for k := 0; k <= K; k++ {
		dp[k] = make([]State, n+1)
	}

	for i := 0; i <= n; i++ {
		dp[0][i] = State{
			score: 0,
			ids:   []int{},
			valid: true,
		}
	}

	better := func(a State, b State) bool {
		if !a.valid {
			return false
		}
		if !b.valid {
			return true
		}

		if a.score != b.score {
			return a.score > b.score
		}

		for i := 0; i < len(a.ids) && i < len(b.ids); i++ {
			if a.ids[i] != b.ids[i] {
				return a.ids[i] < b.ids[i]
			}
		}

		return len(a.ids) < len(b.ids)
	}

	lowerBound := func(length int, target int64) int {
		lo := 0
		hi := length

		for lo < hi {
			mid := lo + (hi-lo)/2

			if ends[mid] >= target {
				hi = mid
			} else {
				lo = mid + 1
			}
		}

		return lo
	}

	for i := 1; i <= n; i++ {
		left := a[i-1][0]
		weight := a[i-1][2]
		originalIndex := int(a[i-1][3])

		p := lowerBound(i-1, left)

		for k := 1; k <= K; k++ {
			dp[k][i] = dp[k][i-1]

			prev := dp[k-1][p]

			if prev.valid {
				ids := append([]int{}, prev.ids...)

				ids = append(ids, originalIndex)

				sort.Ints(ids)

				take := State{
					score: prev.score + weight,
					ids:   ids,
					valid: true,
				}

				if better(take, dp[k][i]) {
					dp[k][i] = take
				}
			}
		}
	}

	var answer State

	for k := 1; k <= K; k++ {
		if better(dp[k][n], answer) {
			answer = dp[k][n]
		}
	}

	return answer.ids
}